### PR TITLE
Signing:  version RepositorySignatures resource in V3 HTTP cache

### LINKED ISSUE #7997: HTTP cache:  RepositoryResources resource is not cached in a versioned way
Although RepositorySignatures resources are [versioned](https://docs.microsoft.com/en-us/nuget/api/repository-signatures-resource#versioning), the cached content is [not versioned](https://github.com/NuGet/NuGet.Client/blob/6454ee58fdd151e0033c3f22df713abc5c1152d6/src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs#L65) in NuGet's HTTP cache.

This means if a 5.0 client populates the cache with a RepositorySignatures/5.0.0 resource where `allRepositorySigned` is `true` and then a 4.7-4.9.3 client uses the cached resource, restore will fail like [this](https://github.com/dotnet/core-eng/issues/5888).

1.  Clear all caches.  (`nuget locals -clear all`)
2.  Restore a package from https://nuget.org using a 5.0 client.
3.  Within 30 minutes, which is how long the HTTP cache is valid, use a 4.9.3 client to restore a different package.

Result:  The 4.9.3 client restore will fail like [this](https://github.com/dotnet/core-eng/issues/5888).

### PR DESCRIPTION
**Pull request by dtivel**
Resolve https://github.com/NuGet/Home/issues/7997.

NuGet 5.0 introduces support for the new [`RepositorySignatures/5.0.0`](https://docs.microsoft.com/en-us/nuget/api/repository-signatures-resource#versioning) V3 service index resource.  This new resource supports the `allRepositorySigned` flag being set to `true`.  Older NuGet clients (i.e.:  4.7.x-4.9.3) do not support this flag properly.

The issue is that NuGet 5.0 currently populates the HTTP cache with a `RepositorySignatures/5.0.0` resource (if found) in a non-versioned cache file (`repository_signatures.dat`).   This means that if a dotnet CLI using NuGet 4.7.x-4.9.3 subsequently uses the cache while its still valid, the client may apply the `allRepositorySigned` `true` flag incorrectly.

The fix is to version the cache file.  A `RepositorySignatures/x.y.z resource` will be cached as `repository_signatures_x.y.z.dat`.  In the scenario above a NuGet 4.7.x-4.9.3 client won't see the 5.0 client's cached file and will correctly request (and subsequently cache) the `RepositorySignatures` resource version it supports even though it will be cached in a non-versioned way.

CC @loic-sharma

### CODE REVIEW

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs [hunk 1]
```diff
@@ -95,5 +98,13 @@ public RepositorySignatureResourceProvider()
 
             return null;
         }
+
+        private static string GenerateCacheKey(ServiceIndexEntry serviceEntry)
+        {
+            var index = serviceEntry.Type.IndexOf('/');
+            var version = serviceEntry.Type.Substring(index + 1).Trim();
+
+            return $"repository_signatures_{version}";
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:107]: 👏

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:107]: Do you have a tracking issue for the same potential problem with the other resources? 
I did some quick looking and for example registration could hit the same problem.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/Providers/PackageDetailsUriResourceV3ProviderTests.cs [hunk 2]
```diff
@@ -11,7 +11,7 @@
 using NuGet.Versioning;
 using Xunit;
 
-namespace NuGet.Protocol.Tests.Providers
+namespace NuGet.Protocol.Providers.Tests
```
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/Providers/PackageDetailsUriResourceV3ProviderTests.cs:14]: Why flip the namespace?

The root namespace is just the name of the Assembly.

**Review by nkolev92 — APPROVED** [review]
A few small concerns. 

It'd be great to have a tracking issue for the other resources. 

While not necessarily critical, I think it's important to fix to eliminate the potential noise from this issue.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/Providers/PackageDetailsUriResourceV3ProviderTests.cs (hunk 2 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Protocol.Tests/Providers/PackageDetailsUriResourceV3ProviderTests.cs:14]: Discussed offline. 
It's fine as it is. We have mixed conventions throughout our test code.

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs (hunk 1 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:107]: https://github.com/NuGet/Home/issues/8003

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs [hunk 3]
```diff
@@ -25,11 +26,11 @@ public RepositorySignatureResourceProvider()
             var serviceIndex = await source.GetResourceAsync<ServiceIndexResourceV3>(token);
             if (serviceIndex != null)
             {
-                var repoSignUrl = serviceIndex.GetServiceEntryUri(ServiceTypes.RepositorySignatures);
+                var serviceEntry = serviceIndex.GetServiceEntries(ServiceTypes.RepositorySignatures).FirstOrDefault();
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs]: What's the value of moving from `GetServiceEntryUri` to `GetServiceEntries`?

Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:29]: >GetServiceEntries [](start = 48, length = 17)

I don't know how common this pattern is, but consider adding a `ServiceIndexResourceV3.GetServiceEntry` method like `GetServiceEntryUri`

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs [hunk 4]
```diff
@@ -95,5 +98,13 @@ public RepositorySignatureResourceProvider()
 
             return null;
         }
+
+        private static string GenerateCacheKey(ServiceIndexEntry serviceEntry)
+        {
+            var index = serviceEntry.Type.IndexOf('/');
+            var version = serviceEntry.Type.Substring(index + 1).Trim();
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:105]: This won't work on service index like this:

```
{
"@type": "RepositorySignatures/Versioned",
"clientVersion": "4.7.0",
 ...
},
{
"@type": "RepositorySignatures/Versioned",
"clientVersion": "5.0.0",
 ...
},
```

I realize that the `RepositorySignatures` resource doesn't support the `RepositorySignatures/Versioned` type, but this feels like an easy trap to fall into in the future if/when we add this type. Also, should we add support for this resource type? We would've been able to enable repository signing for 5.0.0 clients with no client changes had we had this type.

File: test/NuGet.Core.Tests/NuGet.Protocol.Tests/ServiceTypesTests.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    public class ServiceTypesTests
+    {
+        [Fact]
+        public void RepositorySignatures_AllResourcesAreVersioned()
+        {
+            var pattern = new Regex(@"^RepositorySignatures/[4-9]\.\d\.\d$");
+
+            foreach (var repositorySignaturesResource in ServiceTypes.RepositorySignatures)
+            {
+                Assert.True(pattern.IsMatch(repositorySignaturesResource), $"RepositorySignatures resource '{repositorySignaturesResource}' is not versioned with an expected format.  Verify that {nameof(RepositorySignatureResourceProvider)} can handle any new formats.");
```
Comment (loic-sharma) [inline test/NuGet.Core.Tests/NuGet.Protocol.Tests/ServiceTypesTests.cs:20]: 👌

**Review by loic-sharma — APPROVED** [review]

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs (hunk 4 above)
Comment (loic-sharma) [reply src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:105]: Out of curiosity, why doesn't the HTTP cache not use the URL (or some string derived from the URL) as the cache key?

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs (hunk 3 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:29]: It's uncommon now, but if https://github.com/NuGet/Home/issues/8003 uncovers more use cases, I'll add a convenience method.

File: src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs (hunk 4 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Protocol/Providers/RepositorySignatureResourceProvider.cs:105]: Using /Versioned is a good idea, but I think it should be a separate issue/PR.

Not sure why URL's weren't used.  That decision was made >3 years ago.

**Pull request merged**